compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -4078,6 +4080,46 @@ public void visit(BLangBinaryExpr binaryExpr) {
         resultType = types.checkType(binaryExpr, actualType, expType);
     }
 
+    private BType checkAndGetType(BLangExpression expr, SymbolEnv env, BLangBinaryExpr binaryExpr) {
```

Comment:
1. This function looks too nested. Shall we try to early exit when possible.
2. This function only seems to propagate the contextually expected type one level down. We need to support `float x = 1.0 + 2 + 3` too.